openapi/openapi.json

```diff
@@ -1933,10 +1938,21 @@
             "description": "JSON API links object, pointing to a homepage URL for this implementation"
           },
           "link_type": {
-            "$ref": "#/components/schemas/LinkType"
+            "allOf": [
+              {
+                "$ref": "#/components/schemas/LinkType"
+              }
+            ],
```

Comment:
Ugh, I think this switch is going to break things in our internal schema parsing, we never dealt with the `anyOf`/`allOf` types because they were so rare

Reply:
It was in my hacky function though, right?

Reply:
From the tests I see this is not the case with the updated functionality. Great that the tests catch it!

Reply:
I think there might be two hacky functions for this... all needed to be handled in part of #504 but other things have gotten in the way

Reply:
All of the changes revolve around `Enum`s, which are mentioned in the pydantic changes quite heavily... will have to read closely

Reply:
The more I look at this, the more I think this is a pydantic regression? I might raise a few issues with them, 1) enums, 2) "required optional" fields with constrained types, 3) automatically adding `nullable=True` to schema for required optionals

Reply:
I'm convinced of the same, actually, as they already rolled back using `allOf` for enumerations ...

Reply:
:detective:

Reply:
From some further sleuthing, this issue seems to be to do with the interaction between the `Field` description and class docstring. With pydantic 1.6.1, the `Field` description is ignored, and the class docstring is used (and no `anyOf` crops up). With pydantic 1.7.1, it tries to include both the description of the `Field`, and the description of the `Enum` (from its class docstring). See this MWE:

```python
from enum import Enum
from typing import Optional
import json
from pydantic import BaseModel, Field
import pydantic.utils

class MyEnum(Enum):
    """Class docstring. """
    RED = "red"
    BLUE = "blue"

class MyModel(BaseModel):
    colour: Optional[MyEnum] = Field(None, description="Field desc")

if __name__ == "__main__":
    print(pydantic.utils.version_info())
    print(json.dumps(MyModel.schema(), indent=2))
```

Pydantic 1.6.1:
```
$ python mwe.py
             pydantic version: 1.6.1
            pydantic compiled: True
                 install path: /home/mevans/.local/conda/envs/pydantic_17/lib/python3.8/site-packages/pydantic
               python version: 3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0]
                     platform: Linux-5.4.67-1-MANJARO-x86_64-with-glibc2.10
     optional deps. installed: []
{
  "title": "MyModel",
  "type": "object",
  "properties": {
    "colour": {
      "$ref": "#/definitions/MyEnum"
    }
  },
  "definitions": {
    "MyEnum": {
      "title": "MyEnum",
      "description": "Class docstring. ",
      "enum": [
        "red",
        "blue"
      ]
    }
  }
}
```

Pydantic 1.7.1:
```
$ python mwe.py
             pydantic version: 1.7.1
            pydantic compiled: True
                 install path: /home/mevans/.local/conda/envs/pydantic_17/lib/python3.8/site-packages/pydantic
               python version: 3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0]
                     platform: Linux-5.4.67-1-MANJARO-x86_64-with-glibc2.10
     optional deps. installed: []
{
  "title": "MyModel",
  "type": "object",
  "properties": {
    "colour": {
      "description": "Field desc",
      "allOf": [
        {
          "$ref": "#/definitions/MyEnum"
        }
      ]
    }
  },
  "definitions": {
    "MyEnum": {
      "title": "MyEnum",
      "description": "Class docstring. ",
      "enum": [
        "red",
        "blue"
      ]
    }
  }
}

```

And if you delete the field description, with pydantic 1.7.1...:

```shell
$ python mwe.py
             pydantic version: 1.7.1
            pydantic compiled: True
                 install path: /home/mevans/.local/conda/envs/pydantic_17/lib/python3.8/site-packages/pydantic
               python version: 3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0]
                     platform: Linux-5.4.67-1-MANJARO-x86_64-with-glibc2.10
     optional deps. installed: []
{
  "title": "MyModel",
  "type": "object",
  "properties": {
    "colour": {
      "description": "Field desc",
      "allOf": [
        {
          "$ref": "#/definitions/MyEnum"
        }
      ]
    }
  },
  "definitions": {
    "MyEnum": {
      "title": "MyEnum",
      "description": "Class docstring. ",
      "enum": [
        "red",
        "blue"
      ]
    }
  }
}
```
```

Reply:
I'd say the v1.7.1 is an improvement then, since it doesn't throw away the field description information.
And well done in finding this!

However, for this implementation, we should probably just remove the field descriptions then?

Reply:
> I'd say the v1.7.1 is an improvement then, since it doesn't throw away the field description information.
> And well done in finding this!

Well, I think it might still be a regression (or at least worth reporting), unless you know of a reason why it the schema below (which includes both descriptions) is invalid? I should run this through swagger...

```json
{
  "title": "MyModel",
  "type": "object",
  "properties": {
    "colour": {
      "description": "Field desc",
      "$ref": "#/definitions/MyEnum"
    }
  },
  "definitions": {
    "MyEnum": {
      "title": "MyEnum",
      "description": "Class docstring. ",
      "enum": [
        "red",
        "blue"
      ]
    }
  }
}
```

> However, for this implementation, we should probably just remove the field descriptions then?

Probably, especially as it only affects a few fields. I imagine it will be a related problem for the missing titles, which may need to be added explicitly somewhere. This should probably be done in a separate pydantic compatibility PR, right?

Reply:
Ah, didn't follow that it was an invalid schema.

Reply:
I don't think it's invalid as it is (i.e. with `allOf`), but I don't know if including `description` as a key alongside `$ref` is somehow forbidden by the standard (which would be nice and clean, and not break any of our code...)

Reply:
Going to split this PR up, removing the update of `pydantic` (even though that was the original motivation for this, but since the fix to pin to v1.6.1, it's not as crucial anymore).

Reply:
Done. The commits related to updating `pydantic` and the changes introduced by that have been added to the respective dependabot PR.